src/main/java/com/linecorp/armeria/server/ServiceConfig.java

```diff
@@ -0,0 +1,194 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Optional;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.ServiceInvocationContext;
+
+/**
+ * A {@link Service} and its {@link PathMapping} and {@link VirtualHost}.
+ *
+ * @see ServerConfig#serviceConfigs()
+ * @see VirtualHost#serviceConfigs()
+ */
+public final class ServiceConfig {
+
+    private static final Pattern LOGGER_NAME_PATTERN =
+            Pattern.compile("^\\p{javaJavaIdentifierStart}\\p{javaJavaIdentifierPart}*" +
+                            "(?:\\.\\p{javaJavaIdentifierStart}\\p{javaJavaIdentifierPart}*)*$");
+
+    /** Initialized later by {@link VirtualHost} via {@link #build(VirtualHost)}. */
+    private VirtualHost virtualHost;
+
+    private final PathMapping pathMapping;
+    private final String loggerName;
+    private final Service service;
+
+    private String fullLoggerName;
+
+    /**
+     * Creates a new instance.
+     */
+    public ServiceConfig(VirtualHost virtualHost, PathMapping pathMapping, Service service) {
+        this(virtualHost, pathMapping, service, null);
+    }
+
+    /**
+     * Creates a new instance.
+     */
+    public ServiceConfig(VirtualHost virtualHost, PathMapping pathMapping, Service service,
+                         @Nullable String loggerName) {
+        this(pathMapping, service, loggerName);
+        this.virtualHost = requireNonNull(virtualHost, "virtualHost");
+    }
+
+    /**
+     * Creates a new instance.
+     */
+    ServiceConfig(PathMapping pathMapping, Service service, @Nullable String loggerName) {
+        this.pathMapping = requireNonNull(pathMapping, "pathMapping");
+        this.service = requireNonNull(service, "service");
+        this.loggerName = loggerName != null ? validateLoggerName(loggerName, "loggerName")
+                                             : defaultLoggerName(pathMapping);
+    }
+
+    static String validateLoggerName(String value, String propertyName) {
+        requireNonNull(value, propertyName);
+        if (!LOGGER_NAME_PATTERN.matcher(value).matches()) {
+            throw new IllegalArgumentException(propertyName + ": " + value);
+        }
+        return value;
+    }
+
+    private static String defaultLoggerName(PathMapping pathMapping) {
+        final Optional<String> exactOpt = pathMapping.exactPath();
+        if (exactOpt.isPresent()) {
+            return normalizeLoggerName(exactOpt.get());
+        }
+
+        final Optional<String> prefixOpt = pathMapping.prefixPath();
+        if (prefixOpt.isPresent()) {
+            return normalizeLoggerName(prefixOpt.get());
+        }
+
+        return "__UNKNOWN__";
+    }
+
+    private static String normalizeLoggerName(String loggerName) {
```

Comment:
convertPathToLoggerName? Doesn't seem to really normalize, nor is it called when an explicit logger name is set.